File: Comments.py
import datetime

class user(object):
  def __init__(self, name):
    self.name = name
    self._logged_in = False
    self.last_logged_in_at = None
  
  def get_name(self, value = None):
    if(not value):
      return(self.name)
    else:
      self.name = value

  def is_logged_in(self):
    return(self._logged_in)
  
  def get_last_logged_in_at(self):
    return(self.last_logged_in_at)
  
  def log_in(self):
    self._logged_in = True
    self.last_logged_in_at = datetime.datetime.now()
  
  def log_out(self):
    self._logged_in = False
  
  def can_edit(self, comment):
    if (comment.author.name == self.name):
      return True
    else:
      return False
  
  def can_delete(self, comment):
    return False
  
  def to_string(self, comment):
    return("hello by " + comment.author.name) 

class moderator(user):
  def __init__(self, name):
    self.name = name

  def can_delete(self, comment):
    return True

class admin(moderator):
  def __init__(self, name):
      self.name = name

  def can_edit(self, comment):
      return True

class comment(object):
  def __init__(self, author, message, replied_to = None):
    self.author = author
    self.message = message
    self.replied_to = replied_to
    self.created_at = datetime.datetime.now()
    
  def get_author(self, value = None):
     if(not value):
      return(self.author)
     else:
      self.author = value

  def get_message(self, value = None):
    if(not value):
      return(self.message)
    else:
      self.message = value
    
  def get_created_at(self):
    return(self.created_at)
  
  def get_replied_to(self, value = None):
    if(not value):
      return(self.replied_to)
    else:
      self.replied_to = value
  
  def to_string(self):
    if(not self.replied_to):
      return(self.message + " by " + self.author.name)
    else:
      return(self.message + " by " + self.author.name + " (replied to " + self.replied_to.author.name + ")")

user1 = user("Stanley")
user1.log_in()
print(user1.last_logged_in_at)

File: comments.js

class User {
    constructor(name) {
      this.name = name;
      this.lastLoggedInAt = null;
      this.loggedIn = false;
    }
    isLoggedIn() {
      return this.loggedIn;
    }
    getLastLoggedInAt() {
      return this.lastLoggedInAt;
    }
    logIn() {
      this.loggedIn = true;
      this.lastLoggedInAt = new Date();
    }
    logOut() {
      this.loggedIn = false;
    }
    getName() {
      return this.name;
    }
    setName(name) {
      this.name= name;
    }
    canEdit(comment) {
      if (comment.getAuthor().getName() === this.name) {
        return true;
      }
      return false;
    }
    canDelete(comment) {
      return false;
    }
  }
  
  class Moderator extends User {
    constructor(name) {
      super(name);
    }
    canDelete(comment) {
      return true;
    }
  }
  
  class Admin extends Moderator {
    constructor(name) {
      super(name);
    }
    canEdit(comment) {
      return true;
    }
  }
  
  class Comment {
    constructor(author, message, repliedTo) {
      this.createdAt = new Date();
      this.author = author;
      this.message = message;
      this.repliedTo = repliedTo || null;
    }
    getMessage() {
      return this.message;
    }
    setMessage(message) {
      this.message = message;
    }
    getCreatedAt() {
      return this.createdAt;
    }
    getAuthor() {
      return this.author;
    }
    getRepliedTo() {
      return this.repliedTo;
    }
    getString(comment) {
      const authorName = comment.getAuthor().getName();
      if (!comment.getRepliedTo()) return authorName;
      return `${comment.getMessage()} by ${authorName} (replied to ${this.getString(comment.getRepliedTo())})`;
    }
    toString() {
      const authorName = this.getAuthor().getName();
      if (!this.getRepliedTo()) {
        return `${this.message} by ${authorName}`;
      }
      return this.getString(this);
    }
  }
  